File: medium/106. Construct Binary Tree from Inorder and Postorder Traversal/106. Construct Binary Tree from Inorder and Postorder Traversal.py
from typing import List, Optional


# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def buildTree(self, inorder: List[int], postorder: List[int]) -> Optional[TreeNode]:
        if not inorder or not postorder:
            return

        head = postorder.pop()
        position = inorder.index(head)

        right_tree = self.buildTree(inorder[position + 1 :], postorder)

        left_tree = self.buildTree(inorder[:position], postorder)

        return TreeNode(val=head, right=right_tree, left=left_tree)

File: medium/106. Construct Binary Tree from Inorder and Postorder Traversal/106. Construct Binary Tree from Inorder and Postorder Traversal.ts
/**
 * Definition for a binary tree node.
 */

class TreeNode {
    val: number
    left: TreeNode | null
    right: TreeNode | null
    constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
        this.val = (val === undefined ? 0 : val)
        this.left = (left === undefined ? null : left)
        this.right = (right === undefined ? null : right)
    }
}

function buildTree(inorder: number[], postorder: number[]): TreeNode | null {
    if (!inorder.length || !postorder.length)
        return null

    const head = postorder.pop()!
    const indexInInorder = inorder.indexOf(head)

    const right_tree = buildTree(inorder.slice(indexInInorder + 1,), postorder)
    const left_tree = buildTree(inorder.slice(0, indexInInorder), postorder)

    return new TreeNode(head, left_tree, right_tree)
};

export { }

